firedrake/function.py

```diff
@@ -354,13 +355,22 @@ def _interpolate(self, fs, dat, expression, subset):
             raise RuntimeError('Rank mismatch: Expression rank %d, FunctionSpace rank %d'
                                % (expression.rank(), len(fs.ufl_element().value_shape())))
 
-        if expression.value_shape() != fs.ufl_element().value_shape():
+        if expression.ufl_shape != fs.ufl_element().value_shape():
             raise RuntimeError('Shape mismatch: Expression shape %r, FunctionSpace shape %r'
-                               % (expression.value_shape(), fs.ufl_element().value_shape()))
+                               % (expression.ufl_shape, fs.ufl_element().value_shape()))
 
         coords = fs.mesh().coordinates
 
-        if hasattr(expression, "eval"):
+        if not isinstance(expression, expression_t.Expression):
+            kernel, oriented, coefficients = self._interpolate_ufl_expression(expression, to_pts, to_element, fs)
+            args = [kernel, subset or self.cell_set,
+                    dat(op2.WRITE, fs.cell_node_map()[op2.i[0]])]
```

Comment:
I get the wrong values, and they fall into these categories:
* Same values as what I believe to be correct answer, but in the wrong order.
* The sum of all values is the same, but the values themselves are different.
This usually happens with vector-valued elements.